File: lib/commands/devs/shortcut.dart
import 'dart:io';

import 'package:all_exit_codes/all_exit_codes.dart';
import 'package:args/command_runner.dart';

import 'package:xpm/os/shortcut.dart';
import 'package:xpm/utils/logger.dart';

class ShortcutCommand extends Command {
  @override
  final name = "shortcut";
  @override
  final description = "Create a shortcut on the system/desktop";
  @override
  final category = "For developers";

  ShortcutCommand() {
    argParser.addOption("name", abbr: "n", help: "Name of the application", valueHelp: 'name', mandatory: true);

    argParser.addOption("path", abbr: "p", help: "Path of the executable", valueHelp: 'path');

    argParser.addOption("icon", abbr: "i", help: "Name or path of the icon", valueHelp: 'name|path');

    argParser.addOption("description", abbr: "d", help: "Description of the application", valueHelp: 'description');

    argParser.addMultiOption(
      "category",
      abbr: "c",
      help: "Categories, multiple times or once using semicolon",
      valueHelp: 'category[;category2]',
    );

    argParser.addFlag("terminal", abbr: 't', help: 'Run in terminal', defaultsTo: false);

    argParser.addOption(
      "type",
      abbr: 'y',
      help: 'Type of the shortcut',
      valueHelp: 'type',
      allowed: ['Application', 'Link', 'Directory', 'Menu', 'FSDevice', 'FSVolume', 'Location', 'Window'],
      defaultsTo: 'Application',
    );

    argParser.addMultiOption(
      'mime',
      abbr: 'm',
      help: 'MimeTypes, multiple times or once using semicolon',
      valueHelp: 'mime[;mime2]',
    );

    argParser.addFlag("startup", abbr: 'u', help: 'Notify on startup', negatable: true, defaultsTo: true);

    argParser.addFlag("sudo", abbr: 's', help: 'Run as sudo', negatable: true, defaultsTo: true);

    // Remove shortcut flag
    argParser.addFlag("remove", abbr: 'r', help: 'Remove shortcut', negatable: false);
  }

  @override
  void run() async {
    final String name = argResults!['name'];
    final String? executablePath = argResults!['path'];
    final String? icon = argResults!['icon'];
    final String? description = argResults!['description'];
    final List<String> category = argResults!['category'];
    final bool terminal = argResults!['terminal'];
    final String type = argResults!['type'];
    final List<String> mime = argResults!['mime'];
    final bool startup = argResults!['startup'];

    final bool sudo = argResults!['sudo'];
    final bool remove = argResults!['remove'];

    var shortcut = Shortcut(
      name: name,
      executablePath: executablePath,
      icon: icon,
      comment: description,
      categories: category.join(';'),
      terminal: terminal,
      type: type,
      mime: mime.join(';'),
      startup: startup,
      sudo: sudo,
    );

    if (remove) {
      await shortcut.delete();

      Logger.info("Shortcut removed.");
    } else {
      await shortcut.create();

      Logger.info("Shortcut created.");
    }
    exit(success);
  }
}

File: crates/xpm-cli/src/commands/shortcut.rs
//! Desktop shortcut command implementation

use anyhow::Result;
use owo_colors::OwoColorize;
use xpm_core::utils::logger::Logger;

/// Run the shortcut command
#[allow(clippy::too_many_arguments)]
pub async fn run(
    name: &str,
    exec: Option<&str>,
    icon: Option<&str>,
    category: Option<&str>,
    description: Option<&str>,
    terminal: bool,
    shortcut_type: &str,
    mime: Option<&str>,
    startup: bool,
    remove: bool,
) -> Result<()> {
    // Get applications directory
    let apps_dir = dirs::data_dir()
        .map(|d| d.join("applications"))
        .ok_or_else(|| anyhow::anyhow!("Could not determine applications directory"))?;

    let desktop_file = apps_dir.join(format!("{}.desktop", slugify(name)));

    // Handle removal
    if remove {
        if desktop_file.exists() {
            tokio::fs::remove_file(&desktop_file).await?;
            Logger::success(&format!(
                "Removed shortcut: {}",
                desktop_file.display().to_string().green()
            ));

            // Update desktop database
            update_desktop_database(&apps_dir).await;
        } else {
            Logger::warning(&format!("Shortcut not found: {}", name));
        }
        return Ok(());
    }

    // For creation, exec is required
    let exec = exec.ok_or_else(|| anyhow::anyhow!("Executable path is required"))?;

    Logger::info(&format!(
        "Creating desktop shortcut for {}...",
        name.green()
    ));

    tokio::fs::create_dir_all(&apps_dir).await?;

    // Build icon line
    let icon_line = icon
        .map(|i| format!("Icon={}", i))
        .unwrap_or_else(|| "Icon=application-x-executable".to_string());

    // Build category (ensure it ends with semicolon)
    let categories = category
        .map(|c| {
            if c.ends_with(';') {
                c.to_string()
            } else {
                format!("{};", c)
            }
        })
        .unwrap_or_else(|| "Utility;".to_string());

    // Build optional lines
    let comment_line = description
        .map(|d| format!("Comment={}\n", d))
        .unwrap_or_default();

    let mime_line = mime
        .map(|m| {
            if m.ends_with(';') {
                format!("MimeType={}\n", m)
            } else {
                format!("MimeType={};\n", m)
            }
        })
        .unwrap_or_default();

    let content = format!(
        r#"[Desktop Entry]
Type={shortcut_type}
Name={name}
Exec={exec}
{icon_line}
Categories={categories}
Terminal={terminal}
StartupNotify={startup}
{comment_line}{mime_line}"#,
        shortcut_type = shortcut_type,
        name = name,
        exec = exec,
        icon_line = icon_line,
        categories = categories,
        terminal = terminal,
        startup = startup,
        comment_line = comment_line,
        mime_line = mime_line
    );

    // Remove trailing whitespace and extra newlines
    let content = content.trim_end().to_string() + "\n";

    tokio::fs::write(&desktop_file, content).await?;

    // Make executable
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        let mut perms = tokio::fs::metadata(&desktop_file).await?.permissions();
        perms.set_mode(0o755);
        tokio::fs::set_permissions(&desktop_file, perms).await?;
    }

    Logger::success(&format!(
        "Created shortcut at {}",
        desktop_file.display().to_string().green()
    ));

    // Update desktop database
    update_desktop_database(&apps_dir).await;

    Ok(())
}

async fn update_desktop_database(apps_dir: &std::path::Path) {
    if let Ok(status) = tokio::process::Command::new("update-desktop-database")
        .arg(apps_dir)
        .status()
        .await
    {
        if status.success() {
            Logger::info("Updated desktop database");
        }
    }
}

fn slugify(s: &str) -> String {
    s.to_lowercase()
        .chars()
        .map(|c| {
            if c.is_alphanumeric() || c == '-' {
                c
            } else {
                '-'
            }
        })
        .collect::<String>()
        .trim_matches('-')
        .to_string()
}
